File: Stack/20_valid_parenthesis/20.py
import collections

def isValid(s):

    brackets = {
        ')' : '(',
        '}' : '{',
        ']' : '['
    }

    stack = collections.deque()

    for x in s:
        if x not in brackets:
            stack.append(x)
        elif len(stack) != 0 and brackets[x] == stack.pop():
            continue
        else:
            return False
        
    if len(stack) == 0:
        return True
    return False


s = '(()())'
print(f'test case : {s}  result : {isValid(s)}')

s = '{({[]})}}'
print(f'test case : {s}  result : {isValid(s)}')

File: Stack/20_valid_parenthesis/20.rs
use std::collections::HashMap;

fn is_valid(s: &String) -> bool {

    let brackets: HashMap<char, char> = HashMap::from([
        (')', '('),
        ('}', '{'),
        (']', '[')
    ]);
    let string_array: Vec<char> = s.chars().collect();
    let mut stack: Vec<char> = Vec::new();

    for x in string_array {
        if brackets.contains_key(&x) == false {
            stack.push(x);
        } else if (stack.is_empty() == false) && (*brackets.get(&x).unwrap() == stack.pop().unwrap()) {
            continue;
        } else {
            return false;
        }
    }
    
    return stack.is_empty();
}


fn main() {

    let mut s: String = String::from("{{([])[]}}");
    println!("test case : {} result : {} ", s, is_valid(&s));

    s = String::from("{{([][]}}");
    println!("test case : {} result : {} ", s, is_valid(&s));
}